rfcs/00xx_UMA_Unified_Modular_Accelerator_Interface.md

```diff
@@ -0,0 +1,184 @@
+  * Custom primitives can be registered
+  * Lowering from Relay to S-TIR, using TOPI or custom primitives 
+  * Interface for registering accelerator-specific schedules and passes
+  * Execution of UMA schedules and passes on S-TIR
+  * Output:  NS-TIR(including tir.extern calls)
+  * UMALower baseclass (Python only) has to be inherited by accelerator-specific Lower classes (e.g. Accelerator A Lower, etc)
+* UMACodegen
+  * Input: NS-TIR(including tir.extern calls)
+  * Defaults to standard TVM codegen
+  * Intend is to provide a Python interface to insert/emit target code
+  * UMACodegen baseclass has to be inherited by accelerator-specific Codegen classes (e.g. Accelerator A Codegen, etc)
+  * Output: Target .c files
+
+The intention is to use TensorIR and Relax with MetaScheduler for optimization.
+
+
+Abbreviations:
+S-TIR: Schedulable TIR
+NS-TIR: Non-Schedulable TIR
+
+
+### File and class structure and Snippets as example for integration
+UMA provides a mostly python-based API. On the C++ side, new targets are registered using target hooks (RFC #0010). A generic `codegen.cc` handles the calls to the python side.
+```
+.
+├── codegen.cc
+└── targets.cc
+```
+```cpp
+TVM_REGISTER_TARGET_KIND("accelerator_A", kDLCPU)
+    .set_attr<FTVMRelayToTIR>("RelayToTIR", relay::contrib::generic::RelayToTIR("accelerator_A"))
+    .set_attr<FTVMTIRToRuntime>("TIRToRuntime", relay::contrib::generic::accelerator_A::TIRToRuntime);
+
+TVM_REGISTER_TARGET_KIND("accelerator_B", kDLCPU)
+    .set_attr<FTVMRelayToTIR>("RelayToTIR", relay::contrib::generic::RelayToTIR("accelerator_B"));
+    .set_attr<FTVMTIRToRuntime>("TIRToRuntime", relay::contrib::generic::accelerator_B::TIRToRuntime);
+```
+The python API is structured as shown below. The base class `UMABackend` functions as the core API. It uses the API points described in the [flow description](#flow-description) to register elements for the different stages.
+```
+.
+├── backend.py
+├── api
+│   ├── codegen.py
+│   ├── lower.py
+│   ├── partitioner.py
+│   └── utils.py
+├── accelerator_A
+│   ├── backend.py
+│   ├── codegen.py
+│   ├── passes.py
+│   ├── patterns.py
+│   ├── schedules.py
+│   └── strategies.py
+└── accelerator_B
+    └── ...
+```
+A custom accelerator is added by inheriting the `UMABackend`. New elements (e.g., passes, schedules) are added using a registration machanism.
+```python
+"""UMA backend for the UltraTrail accelerator"""
+
+class UltraTrailBackend(UMABackend):
+    def __init__(self):
+        super(UltraTrailBackend, self).__init__()
+
+        # Relay to Relay function registration
+        self._register_pattern("conv1d_relu", conv1d_relu_pattern())
+
+        self._register_relay_pass(1, ConfigGenerator())
+        self._register_relay_pass(2, BufferScopeAnnotator())
+
+        # Relay to TIR function registration
+        self._register_operator_strategy("nn.conv1d", custom_conv1d_strategy, plevel=9)
+
+        self._register_tir_schedule(insert_extern_calls)
```

Comment:
Also, I think it needs explanation/motivation why we need to two hooks ( _register_tir_schedule vs _register_tir_pass)

Reply:
The reason Im asking this the above tvm.tir.Schedule based transform can also be viewed as an implementation detail of S-TIR --> S-TIR. 

Anyway the hook will get S-TIR in the beginning. In an alternate method, one could use : 

_register_tir_pass(STIR_Pass_1) <---- this could/should use tvm.tir.Schedule inside to do the transformation
_register_tir_pass(STIR_Pass_2)
_register_tir_pass(STIR_Pass_3)
_register_tir_pass(NSTIR_Pass_4)
_register_tir_pass(NSTIR_Pass_5)
_register_tir_pass(NSTIR_Pass_6)

Reply:
**TL;DR**
We decided to remove the `_register_tir_schedule` hook for the first version of the UMA API. So transforming S-TIR would work exactly as you described as part of a TIR-pass.

**Background**
Initially it was the plan to create a stable API that would, in the future, also support an easy integration of the MetaScheduler for design space exploration and optimization. This would require a separation of scheduling functions and different kind of transformation passes. For now, we decided to separate the MetaScheduler integration from this RFC and and move it to future work. As a result we removed the additional `_register_tir_schedule` hook.

Reply:
I think is resolved now.